[FLINK-23664][docs] Add docs for KafkaSink (#17058)
apache/flink: docs/content/docs/connectors/datastream/kafka.md

such as time windows or functions with timers, cannot make progress. A single idle Kafka partition causes this behavior.
Consider setting appropriate [idleness timeouts]({{< ref "docs/dev/datastream/event-time/generating_watermarks" >}}#dealing-with-idle-sources) to mitigate this issue.
 

> ## Kafka Sink

Reviewer: Quite a hard start. How about retaining the old intro?

```suggestion
## Kafka Sink

`KafkaSink` allows writing a stream of records to one or more Kafka topics.
```

### Usage

Kafka sink provides a builder class to construct an instance of a KafkaSink. The code snippet below
shows how to write String records to a Kafka topic with a delivery guarantee of at least once.

```java
DataStream<String> stream = ...
        
KafkaSink<String> sink = KafkaSink.<String>builder()
        .setBootstrapServers(brokers)
        .setRecordSerializer(KafkaRecordSerializationSchema.builder()
            .setTopic("topic-name")

>             .setKafkaProducerConfig(producerProperties)

Reviewer: Why is this required?

Author: Hmm, I think it initially was the only way to specify the bootstrapserver. I have a look to make it optional

            .setValueSerializationSchema(new SimpleStringSchema())
            .setDeliveryGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)
            .build()
        )
        .build();
        
stream.sinkTo(sink);
```

The following properties are **required** to build a KafkaSink:

- Bootstrap servers, ```setBootstrapServers(String)```
- Record serializer, ``setRecordSerializer(KafkaRecordSerializationSchema)``
- Kafka producer properties, ```setKafkaProducerConfig(Properties)```
- If you configure the delivery guarantee with ```DeliveryGuarantee.EXACTLY_ONCE``` you also have
  use ```setTransactionalIdPrefix(String)```

### Serializer

> You always need construct a ```KafkaRecordSerializationSchema``` to transform incoming elements from
> the data stream to Kafka producer records.
> We offer a schema builder to provide some common building blocks i.e. key/value serialization, topic
> selection, partitioning.

Reviewer: ```suggestion
You always need supply a ```KafkaRecordSerializationSchema``` to transform incoming elements from
the data stream to Kafka producer records.
Flink offers a schema builder to provide some common building blocks i.e. key/value serialization, topic
selection, partitioning. You can also implement the interface on your own to exert more control.
```

```java
KafkaRecordSerializationSchema.builder()
    .setTopicSelector((element) -> {<your-topic-selection-logic>})
    .setValueSerializationSchema(new SimpleStringSchema())
    .setKeySerializationSchema(new SimpleStringSchema())
    .setPartitioner(new FlinkFixedPartitioner())
    .build();
```

> It is **required** to always set a value serialization method and a topic retrieval method.
> Moreover, it is also possible to use Kafka serializers instead of Flink serializer by using 
> ```setKafkaKeySerializer(Serializer)``` or ```setKafkaValueSerializer(Serializer)```.

Reviewer: ```suggestion
It is **required** to always set a value serialization method and a topic (selection method).
Moreover, it is also possible to use Kafka serializers instead of Flink serializer by using 
```setKafkaKeySerializer(Serializer)``` or ```setKafkaValueSerializer(Serializer)```.
```

### Fault Tolerance

> Overall the ```KafkaSink``` supports all three different ```DeliveryGuarantee```s. For 
> ```DeliveryGuarantee.AT_LEAST_ONCE``` and ```DeliveryGuarantee.EXACTLY_ONCE``` Flink's checkpointing
> must be enabled. By default the ```KafkaSink``` uses ```DeliveryGuarantee.NONE```. Below you can find
> a proper explanation of the different guarantees.

Reviewer: I removed all if we ever add more.
```suggestion
Overall the ```KafkaSink``` supports three different ```DeliveryGuarantee```s. For 
```DeliveryGuarantee.AT_LEAST_ONCE``` and ```DeliveryGuarantee.EXACTLY_ONCE``` Flink's checkpointing
must be enabled. By default the ```KafkaSink``` uses ```DeliveryGuarantee.NONE```. Below you can find
an explanation of the different guarantees.
```

> - ```DeliveryGuarantee.NONE```: does not provide any guarantees: messages may be lost in case of 

Reviewer: ```suggestion
- ```DeliveryGuarantee.NONE``` does not provide any guarantees: messages may be lost in case of 
```

  issues on the Kafka broker and messages may be duplicated in case of a Flink failure.

> - ```DeliveryGuarantee.AT_LEAST_ONCE```: the sink will wait for all outstanding records in the 

Reviewer: ```suggestion
- ```DeliveryGuarantee.AT_LEAST_ONCE```: The sink will wait for all outstanding records in the 
```

  Kafka buffers to be acknowledged by the Kafka producer on a checkpoint. No messages will be 
  lost in case of any issue with the Kafka brokers but messages may be duplicated when Flink 

>   restarts.

Reviewer: ```suggestion
  restarts because Flink reprocesses old input records.
```

> - ```DeliveryGuarantee.EXACTLY_ONCE```: In this mode the KafkaSink will write all messages in a 

Reviewer: ```suggestion
- ```DeliveryGuarantee.EXACTLY_ONCE```: In this mode, the KafkaSink will write all messages in a 
```

  Kafka transaction that will be committed to Kafka on a checkpoint. Thus, if the consumer 
  reads only committed data (see Kafka consumer config isolation.level), no duplicates will be 

>   seen in case of a Flink restart. However, this delays record writing effectively until a 

Reviewer: ```suggestion
  seen in case of a Flink restart. However, this delays record visibility effectively until a 
```
